### PR TITLE
make more standard libraries work with `nimPreviewSlimSystem`

### PR DESCRIPTION
**Pull request by ringabout**

### CODE REVIEW

**Comment by metagn** [comment]
Are there places where we can refactor the use of `assert` into proper errors? 

Also, `assert`/`doAssert` is used in nearly every test/runnable example. Should `std/assertions` be implicitly imported in these? What exactly do we gain from moving `assertions` out of system anyway? It seems like a pretty small module. Is it to support other `assert` implementations?

**Comment by ringabout** [comment]
> Also, assert/doAssert is used in nearly every test/runnable example. Should std/assertions be implicitly imported in these?

Runnable examples shouldn't implicitly import `std/assertions` because users might copy the examples directly, which cannot work without importing `std/assertions`. Same for tests. Too many implicit imports make it hard to run at other places. Cannot just copy the example and run it in other places.

> Are there places where we can refactor the use of assert into proper errors?

I'm sorry, but I don't understand the sentence, could you please elaborate more?

> What exactly do we gain from moving assertions out of system anyway? It seems like a pretty small module. Is it to support other assert implementations?

Perhaps, it is still not used in every module. And as a system programming language, the less implicit imports, the better.

**Comment by ringabout** [comment]
I suppose the `nimPreviewSlimSystem` should be firstly applied to embedded systems which requires bare imports and rebuilds everything.

**Comment by metagn** [comment]
> could you please elaborate more?

For example:

https://github.com/nim-lang/Nim/blob/e0be71cd046e458e87498d2dd9cb69babe4f17e7/lib/impure/nre.nim#L488-L489

A user can give an `endpos` that is outside of the string bounds. If they do this they will get an AssertionDefect saying `strlen <= str.len` instead of a ValueError or RangeDefect or whatever. In this case `assert` is a "lazy" error and should be refactored into a more proper one like `rangeCheck strlen <= str.len`. Sometimes this might elide the need to import `assertions`.

> as a system programming language, the less implicit imports, the better.

Assertions seem to me like one of the less harmful things to keep in system. IMO it's worth not spamming `import std/assertions` everywhere.

At least just having `assert` & `doAssert` in system is fine IMO, stuff like `doAssertRaises` are a bit much and can stay in an external module (probably still `assertions`). `failedAssertImpl` and `raiseAssert` shouldn't be exposed either. Sorry to hijack this PR to say this, I will make a separate proposal if I decide it's a good idea.

Edit: Done in #20354

File: tests/stdlib/tjsontestsuite.nim [hunk 1]
```diff
@@ -19,19 +19,19 @@ let parsing_testdata = {
   "i_string_incomplete_surrogates_escape_valid": """["\uD800\uD800\n"]""",
   "i_string_inverted_surrogates_U+1D11E": """["\uDd1e\uD834"]""",
   "i_string_lone_second_surrogate": """["\uDFAA"]""",
-  "i_string_not_in_unicode_range": """["�"]""",
-  "i_string_truncated-utf-8": """["��"]""",
```
Comment (Araq) [inline tests/stdlib/tjsontestsuite.nim]: What's up with these?

Comment (ringabout) [reply tests/stdlib/tjsontestsuite.nim]: Sorry, fixed. Not sure what vscode did to this file.

Comment (ringabout) [reply tests/stdlib/tjsontestsuite.nim]: I didn't know GitHub cannot show my changes properly

https://github.com/nim-lang/Nim/pull/20343/commits/d93fb0c96aa73cdc14ec1585680759477d7c50db

At least this file is not changed anymore.

image

**Comment by ringabout** [comment]
Unrelated CI failure

**Pull request merged**